Source/Core/Core/IOS/USB/Emulated/Infinity.h

```diff
@@ -0,0 +1,118 @@
+  std::vector<InterfaceDescriptor> GetInterfaces(u8 config) const override;
+  std::vector<EndpointDescriptor> GetEndpoints(u8 config, u8 interface, u8 alt) const override;
+  bool Attach() override;
+  bool AttachAndChangeInterface(u8 interface) override;
+  int CancelTransfer(u8 endpoint) override;
+  int ChangeInterface(u8 interface) override;
+  int GetNumberOfAltSettings(u8 interface) override;
+  int SetAltSetting(u8 alt_setting) override;
+  int SubmitTransfer(std::unique_ptr<CtrlMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<BulkMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<IntrMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<IsoMessage> message) override;
+  void ScheduleTransfer(std::unique_ptr<TransferCommand> command, const std::array<u8, 32>& data,
+                        s32 expected_count, u64 expected_time_us);
+
+private:
+  Kernel& m_ios;
+  u16 m_vid = 0;
+  u16 m_pid = 0;
+  u8 m_active_interface = 0;
+  bool m_device_attached = false;
+  DeviceDescriptor m_device_descriptor;
+  std::vector<ConfigDescriptor> m_config_descriptor;
+  std::vector<InterfaceDescriptor> m_interface_descriptor;
+  std::vector<EndpointDescriptor> m_endpoint_descriptor;
+  std::queue<std::array<u8, 32>> m_queries;
+  std::queue<std::unique_ptr<IntrMessage>> m_response_list;
+};
+
+struct InfinityFigure final
+{
+  File::IOFile inf_file;
+  std::array<u8, 0x14 * 0x10> data{};
+  bool present = false;
+  u8 order_added = 255;
+  void Save();
+};
+
+class InfinityBase final
+{
+public:
+  bool HasFigureBeenAddedRemoved();
+  std::array<u8, 32> GetAddedRemovedResponse();
+  void GetBlankResponse(u8 sequence, u8* reply_buf);
+  void GetPresentFigures(u8 sequence, u8* reply_buf);
+  void GetFigureIdentifier(u8 fig_num, u8 sequence, u8* reply_buf);
+  void QueryBlock(u8 fig_num, u8 block, u8* reply_buf, u8 sequence);
+  void WriteBlock(u8 fig_num, u8 block, const u8* to_write_buf, u8* reply_buf, u8 sequence);
+  void RemoveFigure(u8 position);
+  std::string LoadFigure(u8* buf, File::IOFile in_file, u8 position);
+  bool CreateFigure(const std::string& file_path, u16 character);
+  void CopyBlockToFile(std::array<u8, 16> data, u8* file, u8 position);
+  std::string FindFigure(u16 character);
+  void GenerateSeed(u32 seed);
+  u32 GetNext();
+  u64 Scramble(u32 num_to_scramble, u32 garbage);
+  u32 Descramble(u64 num_to_descramble);
+  u8 GenerateChecksum(u8* data, int num_of_bytes);
+
+protected:
+  std::mutex infinity_mutex;
+  InfinityFigure m_player_one;
+  InfinityFigure m_player_two;
+  InfinityFigure m_player_one_ability;
+  InfinityFigure m_player_two_ability;
+  InfinityFigure m_hexagon;
+
+private:
+  u32 _a;
+  u32 _b;
+  u32 _c;
+  u32 _d;
+  u32 RotateLeft(u32 x, u8 d);
+  u8 m_figure_order = 0;
+  std::queue<std::array<u8, 32>> m_figure_added_removed_response;
+
+  const std::vector<u8> SHA1_CONSTANT = {0xAF, 0x62, 0xD2, 0xEC, 0x04, 0x91, 0x96, 0x8C,
```

Comment:
This should be a static constexpr array. If it needs to be added to another container type, then the initialization of that specific container can be

```cpp
Container container(SHA1_CONSTANT.begin(), SHA1_CONSTANT.end());
```